src/cudd/adapter.h

```diff
@@ -78,6 +78,289 @@ class cudd_adapter
+  std::vector<int> _permute_relprev;
+
+  // Init and Deinit
+public:
+  cudd_bdd_adapter(int varcount)
+    : cudd_adapter(varcount, 0)
+  { // Disable dynamic ordering
+    if (!enable_reordering) { _mgr.AutodynDisable(); }
+
+    _latest_build = bot();
+  }
+
+  // BDD Operations
+public:
+  inline ADD
+  top()
+  {
+    return _mgr.addOne();
+  }
+
+  inline ADD
+  bot()
+  {
+    return _mgr.addZero();
+  }
+
+  inline ADD
+  ithvar(int i)
+  {
+    return _mgr.addVar(i);
+  }
+
+  inline ADD
+  nithvar(int i)
+  {
+    return ~_mgr.addVar(i);
+  }
+
+  template <typename IT>
+  inline ADD
+  cube(IT rbegin, IT rend)
+  {
+    ADD res = top();
+    while (rbegin != rend) { res = _mgr.addVar(*(rbegin++)).Ite(res, bot()); }
+    return res;
+  }
+
+  inline ADD
+  cube(const std::function<bool(int)>& pred)
+  {
+    ADD res = top();
+    for (int i = _varcount - 1; 0 <= i; --i) {
+      if (pred(i)) { res = _mgr.addVar(i).Ite(res, bot()); }
+    }
+    return res;
+  }
+
+  inline ADD
+  apply_and(const ADD& f, const ADD& g)
+  {
+    return f & g;
+  }
+
+  inline ADD
+  apply_or(const ADD& f, const ADD& g)
+  {
+    return f.Or(g);
+  }
+
+  inline ADD
+  apply_diff(const ADD& f, const ADD& g)
+  {
+    return f.Diff(g);
+  }
+
+  inline ADD
+  apply_imp(const ADD& f, const ADD& g)
+  {
+    return (~f).Or(g);
+  }
+
+  inline ADD
+  apply_xor(const ADD& f, const ADD& g)
+  {
+    return f.Xor(g);
+  }
+
+  inline ADD
+  apply_xnor(const ADD& f, const ADD& g)
+  {
+    return f.Xnor(g);
+  }
+
+  inline ADD
+  ite(const ADD& f, const ADD& g, const ADD& h)
+  {
+    return f.Ite(g, h);
+  }
+
+  template <typename IT>
+  inline ADD
+  extend(const ADD& f, IT /*begin*/, IT /*end*/)
+  {
+    return f;
+  }
+
+  inline ADD
+  exists(const ADD& f, int i)
+  {
+    return f.ExistAbstract(_mgr.addVar(i));
+  }
+
+  inline ADD
+  exists(const ADD& f, const std::function<bool(int)>& pred)
+  {
+    return f.ExistAbstract(cube(pred));
+  }
+
+  template <typename IT>
+  inline ADD
+  exists(const ADD& f, IT rbegin, IT rend)
+  {
+    return f.ExistAbstract(cube(rbegin, rend));
+  }
+
+  inline ADD
+  forall(const ADD& f, int i)
+  {
+    return f.UnivAbstract(_mgr.addVar(i));
+  }
+
+  inline ADD
+  forall(const ADD& f, const std::function<bool(int)>& pred)
+  {
+    return f.UnivAbstract(cube(pred));
+  }
+
+  template <typename IT>
+  inline ADD
+  forall(const ADD& f, IT rbegin, IT rend)
+  {
+    return f.UnivAbstract(cube(rbegin, rend));
+  }
+
+  inline ADD
+  relnext(const ADD& states, const ADD& rel, const ADD& /*rel_support*/)
+  {
+
+    if (_permute_relnext.empty()) {
+      _vars_relnext = cube([](int x) { return x % 2 == 0; });
+
+      _permute_relnext.reserve(_varcount);
+      for (int x = 0; x < _varcount; ++x) { _permute_relnext.push_back(x & ~1); }
+    }
+
+    // FIXME: there is no specialized `ADD::AndAbstract` in CUDD yet
+    return (states & rel).ExistAbstract(_vars_relnext).Permute(_permute_relnext.data());
```

Comment:
Fun fact (not that you need to change anything). But, CUDD does surprisingly well despite of this.